jax/scipy/sparse/linalg.py

```diff
@@ -167,10 +183,254 @@ def cg(A, b, x0=None, *, tol=1e-5, atol=0.0, maxiter=None, M=None):
 
   cg_solve = partial(
       _cg_solve, x0=x0, tol=tol, atol=atol, maxiter=maxiter, M=M)
+
   # real-valued positive-definite linear operators are symmetric
-  real_valued = lambda x: not issubclass(x.dtype.type, np.complexfloating)
+  def real_valued(x):
+    return not issubclass(x.dtype.dtype, np.complexfloating)
   symmetric = all(map(real_valued, tree_leaves(b)))
   x = lax.custom_linear_solve(
       A, b, solve=cg_solve, transpose_solve=cg_solve, symmetric=symmetric)
   info = None  # TODO(shoyer): return the real iteration count here
   return x, info
+
+
+def _safe_normalize(x, return_norm=False, thresh=None):
+  norm = jnp.sqrt(_vdot_tree(x, x, assume_real=False))
+  if thresh is None:
+    thresh = jnp.finfo(x.dtype).eps
+  thresh = thresh.astype(norm.dtype)
+
+  normalized_x, norm = lax.cond(
+    norm > thresh,
+    lambda y: (_div(y, norm), norm),
+    lambda y: (_mul(0.*y, thresh), thresh),  # To get the dtype right
```

Comment:
`0.*y` will fail if `y` is a pytree.
```suggestion
    lambda y: tree_map(jnp.zeros_like, y),  # To get the dtype right
```